sim/src/main/cc/bridges/TracerVModule.cc

```diff
@@ -0,0 +1,367 @@
+                << std::endl;
+    }
+
+    return was_done;
+  }
+
+  using expected_t = std::pair<uint64_t, std::vector<uint64_t>>;
+
+  std::vector<expected_t> expected_pair;
+
+  void write_iaddr(const unsigned i, const uint64_t x) {
+    peek_poke.poke(namei(i), x, true);
+
+    // as a simplification, also write a modified version of
+    // the PC value as the instruction
+    peek_poke.poke(nameinsn(i), fold(x), true);
+  }
+
+  void write_valid(const unsigned i, const bool x) {
+    peek_poke.poke(namev(i), x, true);
+  }
+
+  int simulation_run() override {
+    tracerv.init();
+
+    // write the header to our expected test output file
+    tracerv.write_header(expected);
+
+    // Reset the DUT.
+    peek_poke.poke("reset", 1, /*blocking=*/true);
+    step(1, /*blocking=*/true);
+    peek_poke.poke("reset", 0, /*blocking=*/true);
+    step(1, /*blocking=*/true);
+
+    // the value of the first cycle as returned from TracerV
+    const uint64_t cycle_offset = 3;
+
+    // modified as we go
+    uint64_t e_cycle = cycle_offset;
+
+    const unsigned tracerv_width = tracerv.max_core_ipc;
+
+    // load MMIO and capture expected outputs
+    auto load = [&](std::vector<uint64_t> iad, std::vector<bool> bit) {
+      std::vector<uint64_t> valid_i;
+      assert(iad.size() == bit.size());
+      for (unsigned i = 0; i < iad.size(); i++) {
+        // std::cout << "loading " << i << " with " << iad[i] << "," << bit[i]
+        // << std::endl;
+
+        write_iaddr(i, iad[i]);
+        write_valid(i, bit[i]);
+
+        // calculate what TraverV should output, and save it
+        if (bit[i]) {
+          valid_i.emplace_back(iad[i]);
+        }
+      }
+
+      // place instructions onto the vector 7 at a time
+      for (size_t i = 0; i < valid_i.size(); i += 7) {
+        auto last = std::min(valid_i.size(), i + 7);
+        std::vector<uint64_t> chunk =
+            std::vector<uint64_t>(valid_i.begin() + i, valid_i.begin() + last);
+        expected_pair.emplace_back(std::make_pair(e_cycle, chunk));
+      }
+
+      e_cycle++;
+    };
+
+    // loop over tests. choose random valids with a simple pattern of iaddr
+    // load into MMIO, and tick the system
+    for (unsigned test_step = 0; test_step < get_total_trace_tests();
+         test_step++) {
+      const uint64_t pull = rand_next(tracerv_width + 1);
+
+      auto pull_iaddr = get_iaddrs(test_step, tracerv_width);
+      auto pull_bits = get_contiguous(pull, tracerv_width);
+
+      load(pull_iaddr, pull_bits);
+      steps(1);
+    }
+
+    const auto &[final_iaddr, final_valid] = get_final_values(tracerv_width);
+
+    // load final values (which are not valid and thus not checked)
+    load(final_iaddr, final_valid);
+
+    tracerv.flush();
+
+    steps(100);
+
+    // if testing a trigger mode, the output need to be filtered
+    filter_expected_test();
+
+    // write out a file which contains the expected output
+    write_expected_file();
```

Comment:
If my understanding is correct, I would rather create a copy of the filtered list instead of modifying it in-place. Should help with readability a lot: ```write_expected_file(filter_expected_test())```